apps/docs/utils/parse-markdown.tsx

```diff
@@ -0,0 +1,256 @@
+import { assert, assertExists } from '@tldraw/utils'
+import GithubSlugger from 'github-slugger'
+import { Nodes, Root } from 'mdast'
+import { MdxJsxFlowElement, MdxJsxTextElement } from 'mdast-util-mdx'
+import { remark } from 'remark'
+import remarkMdx from 'remark-mdx'
+
+// Add nodes to mdast content.
+declare module 'mdast' {
+	interface BlockContentMap {
+		/**
+		 * MDX JSX element node, occurring in flow (block).
+		 */
+		mdxJsxFlowElement: MdxJsxFlowElement
+	}
+
+	interface PhrasingContentMap {
+		/**
+		 * MDX JSX element node, occurring in text (phrasing).
+		 */
+		mdxJsxTextElement: MdxJsxTextElement
+	}
+
+	interface RootContentMap {
+		/**
+		 * MDX JSX element node, occurring in flow (block).
+		 */
+		mdxJsxFlowElement: MdxJsxFlowElement
+		/**
+		 * MDX JSX element node, occurring in text (phrasing).
+		 */
+		mdxJsxTextElement: MdxJsxTextElement
+	}
+}
+
+export interface ParsedHeading {
+	level: number
+	title: string
+	slug: string
+	contentText: string
+	isInherited: boolean
+}
+
+/**
+ * Walk the MDX AST of `content` and extract headings & their (plain text) content as we go.
+ *
+ * This returns an array of heading objects containing the heading info, and the text below it
+ * (before the next heading). It also returns the initial content text before the first heading.
+ *
+ * Heading extraction should be reliable, but content extraction is pretty best-effort. It's useful
+ * for a search index (which is what we use it for!), but might not be a perfect representation of
+ * everything on the page.
+ */
+export function parseHeadings(content: string): {
+	headings: ParsedHeading[]
+	initialContentText: string
+} {
+	const headings: Omit<ParsedHeading, 'slug'>[] = [
+		{
+			level: 0,
+			title: '',
+			contentText: '',
+			isInherited: false,
+		},
+	]
+
+	remark()
+		.use(remarkMdx)
+		.use(() => (tree: Root) => {
+			interface State {
+				isInHeading: boolean
+				isInherited: boolean
+			}
+
+			function visitChildren(node: Nodes, state: State) {
+				if ('children' in node) {
+					for (const child of node.children) {
+						visit(child, state)
+					}
+				}
+			}
+
+			function heading() {
+				const currentHeading = headings[headings.length - 1]
+				return currentHeading
+			}
+
+			function visit(node: Nodes, state: State) {
+				switch (node.type) {
+					case 'heading': {
+						const { depth } = node
+
+						headings.push({
+							level: depth,
+							title: '',
+							contentText: '',
+							isInherited: state.isInherited,
+						})
+
+						visitChildren(node, { ...state, isInHeading: true })
+
+						break
+					}
+
+					case 'text':
+					case 'inlineCode':
+						if (state.isInHeading) {
+							heading().title += node.value
+						} else {
+							heading().contentText += node.value
+						}
+
+						break
+
+					case 'blockquote':
+					case 'paragraph':
+						assert(!state.isInHeading)
+						if (!heading().contentText.endsWith('\n')) {
+							heading().contentText += '\n'
```

Comment:
we need to add some sort of space here so that `foo\nbar.\nbaz` doesn't become `foobar.Baz`. A newline seems to preserve meaning better than a plain space.